state/watcher.go

```diff
@@ -2102,8 +1853,8 @@ func watchInstanceCharmProfileCompatibilityData(backend modelBackend, watchDocId
 	}
 	extract := func(data map[string]interface{}) (string, error) {
 		if value, ok := data["charmurl"]; ok {
-			if url, ok := value.(*charm.URL); ok {
-				return url.String(), nil
+			if url, ok := value.(string); ok {
```

Comment:
This fixes a bug where the uniter was printing an error of "invalid applicationDoc type"

Reply:
The type of the object should actually be a map[string]interface{} since we should have a Document type here, and we deserialized into a map[string]interface{} so nested documents should have that same type.

Reply:
see https://github.com/juju/juju/pull/10081